### ISSUE TITLE
fast-attention: train well but predict badly

### ISSUE THREAD

**Issue by 2020zyc** [google-research/google-research#508]
1. I use the transformer model in models/official/nlp/transformer/transformer.py to train a seq2seq model, replacing the built-in keras attention with performer fast-attention(tensorflow version),  for example, the decoder stack is as follows
![image](https://user-images.githubusercontent.com/7539692/103272848-d76e9e00-49f8-11eb-9594-b0f57114e67a.png)
> It seams images can't be displayed: 
```
    # DecoderStack
    def build(self, input_shape):
        """Builds the decoder stack."""
        params = self.params
        for _ in range(params["num_hidden_layers"]):
            if not params["use_fast_attn"]:  # ==========set to True to use fast_attention=============
                self_attention_layer = attention_layer.SelfAttention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"])
                enc_dec_attention_layer = attention_layer.Attention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"])
            else:
                self_attention_layer = fast_attention.SelfAttention(  # ==========use fast_attention=============
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"],
                    kernel_transformation=fast_attention.softmax_kernel_transformation,
                    causal=True,
                    projection_matrix_type=1,
                    nb_random_features=nb_random_features,  # =============100(model hidden size is 512)============
                )
                enc_dec_attention_layer = fast_attention.Attention(  # ==========use fast_attention=============
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"],
                    kernel_transformation=fast_attention.softmax_kernel_transformation,
                    causal=False,
                    projection_matrix_type=1,
                    nb_random_features=nb_random_features,
                )
            feed_forward_network = ffn_layer.FeedForwardNetwork(
                params["hidden_size"], params["filter_size"], params["relu_dropout"])

            self.layers.append([
                PrePostProcessingWrapper(self_attention_layer, params),  # x->LNorm(x)->{layer(x)=y}->Dropout(y)->y+x
                PrePostProcessingWrapper(enc_dec_attention_layer, params),
                PrePostProcessingWrapper(feed_forward_network, params)
            ])
        self.output_normalization = tf.keras.layers.LayerNormalization(
            epsilon=1e-6, dtype="float32")
        super(DecoderStack, self).build(input_shape)
```

2. The training stage is perfect, with good decoder outputs almost the same as the targets. An example is, if the target is **"He is playing basktball in the basketball gym"**, the output could be the same.

3. But when doing inference, the decoder outputs are very bad. The output for the above example now is **"He is playing is playing is playing is playing is playing </s>"**, even more worse with **"He He He He He He He He He He He He </s>"**.

4. I just change the mode from "train" to "predict", and use beam search. Whether setting beam_size=1 or not, the output is the same and bad.

5. When doing inference, I check the **decoder attention output(layer0)** in "fast_attention.py>Attention>call()", the states between beams of an sample are almost the same in step 2. 
![image](https://user-images.githubusercontent.com/7539692/103276792-7b107c00-4a02-11eb-8604-f9f72ae47bee.png)
![image](https://user-images.githubusercontent.com/7539692/103276588-ec9bfa80-4a01-11eb-9c57-c4edb4565d63.png)

> It seams images can't be displayed: 
```      
        if cache is not None:
            # Combine cached keys and values with new keys and values.
            if decode_loop_step is not None:
                cache_k_shape = cache["k"].shape.as_list()
                indices = tf.reshape(
                    tf.one_hot(decode_loop_step, cache_k_shape[1], dtype=key.dtype),
                    [1, cache_k_shape[1], 1, 1])
                key = cache["k"] + key * indices
                cache_v_shape = cache["v"].shape.as_list()
                indices = tf.reshape(
                    tf.one_hot(decode_loop_step, cache_v_shape[1], dtype=value.dtype),
                    [1, cache_v_shape[1], 1, 1])
                value = cache["v"] + value * indices
        else:
            key = tf.concat([tf.cast(cache["k"], key.dtype), key], axis=1)
            value = tf.concat([tf.cast(cache["v"], value.dtype), value], axis=1)
    
            # Update cache
            cache["k"] = key
            cache["v"] = value

        attention_output = favor_attention(query, key, value,
                                           self.kernel_transformation, self.causal,
                                           bias, projection_matrix)
        
        # the attention_output is: 
        attention_output = tf.Tensor(
        [[-9.661879   2.9591107  1.3627272 ... -5.7556434 -4.247857   5.5499997]
         [-9.661878   2.9591098  1.3627269 ... -5.7556424 -4.247855   5.55     ]
         [-9.661879   2.9591098  1.3627272 ... -5.755643  -4.247857   5.5499997]
         [-9.661878   2.9591103  1.3627266 ... -5.755642  -4.2478557  5.549999 ]
         [-9.661879   2.9591098  1.3627272 ... -5.7556434 -4.247856   5.5499997]], shape=(5, 512), dtype=float32)
```

6. Using the raw code and the built-in keras attention, everying is good.

**Comment by 2020zyc** [comment]
any help, thanks.

**Comment by 2020zyc** [comment]
Currently the problem is:

While inferring, the first token can be generated correctly, but it starts generating repeatedly from the second step. So the output looks like "PAD A A A A A A A...". (A is expected; PAD is the start token of sentence of decoder.)

After the first token has been generated, I remove the PAD-token, then the second token can be correct, but afterwards it repeats again. The output looks like "A B C B C B C B C..." or "A B B B B B B B B..." (A B C are all expected)

**Comment by johnght** [comment]
Just another fast attention user. Haven't study your model.
From my observation, the attention_dropout and the training are not yet supported in the build and call function correspondingly.
In my model, the validation loss is higher than the training loss. Hence, overfitting.
Is this the reason behind your 'train well predict bad' scenario? If yes, place a separate Dropout layer after this fast Attention layer.
ps: tensorflow supports embedded dropout badly. e.g. LSTM(dropout=...) memory leakage since 2.0 not fixed until recently >2.3?

**Comment by 2020zyc** [comment]
> Just another fast attention user. Haven't studied your model.
> From my observation, the attention_dropout and the training are not yet supported in the build and call function correspondingly.
> In my model, the validation loss is higher than the training loss. Hence, overfitting.
> Is this the reason behind your 'train well predict bad' scenario? If yes, place a separate Dropout layer after this fast Attention layer.
> ps: tensorflow supports embedded dropout badly. e.g. LSTM(dropout=...) memory leakage since 2.0 not fixed until recently >2.3?

Thanks, I just tried it and indeed didn't delve into the implementation of fast-attention, sorry. 

Out of confidence in the author, I didn't think it was an simple overfitting problem due to the invalid dropout, but a complex one.

I will read the code in detail. 

Thanks again.

**Comment by 2020zyc** [comment]
> Just another fast attention user. Haven't studied your model.
> From my observation, the attention_dropout and the training are not yet supported in the build and call function correspondingly.
> In my model, the validation loss is higher than the training loss. Hence, overfitting.
> Is this the reason behind your 'train well predict bad' scenario? If yes, place a separate Dropout layer after this fast Attention layer.
> ps: tensorflow supports embedded dropout badly. e.g. LSTM(dropout=...) memory leakage since 2.0 not fixed until recently >2.3?

I check the code and it's true for "the attention_dropout and the training are not yet supported in the build".

For "If yes, place a separate Dropout layer after this fast Attention layer.", each layer has been wrapped by the PrePostProcessingWrapper layer, which include the layer norm and dropout:
```
# code in DecodeStack/build()
...
        for _ in range(params["num_hidden_layers"]):
            if not params["use_fast_attn"]:
                self_attention_layer = attention_layer.SelfAttention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"])
                enc_dec_attention_layer = attention_layer.Attention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"])
            else:
                self_attention_layer = fast_attention.SelfAttention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"],
                    kernel_transformation=fast_attention.softmax_kernel_transformation,
                    causal=True,
                    projection_matrix_type=1,
                    nb_random_features=nb_random_features,
                )
                enc_dec_attention_layer = fast_attention.Attention(
                    params["hidden_size"], params["num_heads"],
                    params["attention_dropout"],
                    kernel_transformation=fast_attention.softmax_kernel_transformation,
                    causal=False,
                    projection_matrix_type=1,
                    nb_random_features=nb_random_features,
                )
            feed_forward_network = ffn_layer.FeedForwardNetwork(
                params["hidden_size"], params["filter_size"], params["relu_dropout"])
            # ==================wrap each layer===================
            self.layers.append([
                PrePostProcessingWrapper(self_attention_layer, params),  # x->LNorm(x)->{layer(x)=y}->Dropout(y)->y+x
                PrePostProcessingWrapper(enc_dec_attention_layer, params),
                PrePostProcessingWrapper(feed_forward_network, params)
            ])
...
# And the PrePostProcessingWrapper
class PrePostProcessingWrapper(tf.keras.layers.Layer):
    """Wrapper class that applies layer pre-processing and post-processing."""
    def __init__(self, layer, params):
        super(PrePostProcessingWrapper, self).__init__()
        self.layer = layer
        self.params = params
        self.postprocess_dropout = params["layer_postprocess_dropout"]

    def build(self, input_shape):
        # Create normalization layer
        self.layer_norm = tf.keras.layers.LayerNormalization(
            epsilon=1e-6, dtype="float32")
        super(PrePostProcessingWrapper, self).build(input_shape)

    def get_config(self):
        return {
            "params": self.params,
        }
    def call(self, x, *args, **kwargs):
        """Calls wrapped layer with same parameters."""
        # Preprocessing: apply layer normalization
        training = kwargs["training"]

        # axis=-1
        y = self.layer_norm(x)  #  ============using pre layer norm==============

        # Get layer output
        y = self.layer(y, *args, **kwargs)

        # Postprocessing: apply dropout and residual connection
        if training:
            y = tf.nn.dropout(y, rate=self.postprocess_dropout)  # ========using dropout when training=========
        return x + y  #
```

**Comment by 2020zyc** [comment]
I train the transformer model with the built-in attention. The predictions are good with results like "A B C D E F G...".

Then I almost directly replace the built-in attention with the fast one(TF performer). Without any re-training, the predictions become bad with results containing repetive words like "A B B B..."

Maybe the TF-version performer has bug. I'll keep going through the code.

**Comment by johnght** [comment]
Your inference output seems stuttering in the early stage. Then it must be related to the inference cycle. Say the proper use of the 'cache' parameter in the function call() because, unlike ordinary RNN with self-attention, fast attention relies on manual relaying of internal states. If it's still not working, then there maybe programming bugs in the inference logic.
